File: android/app/src/main/java/com/focusapp/AppMonitorService.java
package com.focusapp;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.Service;
import android.app.usage.UsageStats;
import android.app.usage.UsageStatsManager;
import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.os.IBinder;
import android.os.Process;
import android.util.Log;

import androidx.annotation.Nullable;
import androidx.core.app.NotificationCompat;

import java.util.List;
import java.util.SortedMap;
import java.util.TreeMap;

public class AppMonitorService extends Service {
    private static final String TAG = "AppMonitorService";
    private static final String CHANNEL_ID = "monitor_channel";
    private static final int NOTIFICATION_ID = 1;

    private volatile boolean isRunning = false;

    @Override
    public void onCreate() {
        super.onCreate();
        createNotificationChannel();
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        Log.d(TAG, "onStartCommand: service starting…");

        // 1) Build and post the foreground notification
        Notification notification = new NotificationCompat.Builder(this, CHANNEL_ID)
                .setContentTitle("App Monitor")
                .setContentText("Monitoring foreground apps…")
                .setSmallIcon(R.drawable.ic_lock)     // replace with your drawable
                .setOngoing(true)
                .build();

        startForeground(NOTIFICATION_ID, notification);

        // 2) Start the monitoring thread if not already running
        if (!isRunning) {
            isRunning = true;
            startMonitoringLoop();
        }

        // If killed, restart with the last Intent
        return START_STICKY;
    }

    private void startMonitoringLoop() {
        new Thread(() -> {
            UsageStatsManager usm = (UsageStatsManager) getSystemService(Context.USAGE_STATS_SERVICE);
            while (isRunning) {
                try {
                    Thread.sleep(1000);

                long currentTime = System.currentTimeMillis();

                List<UsageStats> stats = usageStatsManager.queryUsageStats(
                    UsageStatsManager.INTERVAL_DAILY,
                    currentTime - 1000 * 1000,
                    currentTime);

                    if (stats != null && !stats.isEmpty()) {
                        SortedMap<Long, UsageStats> sorted = new TreeMap<>();
                        for (UsageStats entry : stats) {
                            sorted.put(entry.getLastTimeUsed(), entry);
                        }
                        String topPackage = sorted.get(sorted.lastKey()).getPackageName();
                        Log.d(TAG, "Foreground app: " + topPackage);

                        // TODO: if topPackage is in your “blocked” list,
                        //       launch your lock-screen activity here
                        //
                        // Intent lockIntent = new Intent(this, LockScreenActivity.class);
                        // lockIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TOP);
                        // startActivity(lockIntent);
                    }
                } catch (InterruptedException e) {
                    Log.e(TAG, "Monitoring thread interrupted", e);
                }
            }
        }).start();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        isRunning = false;
        Log.d(TAG, "onDestroy: service stopped");
    }

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        // We don't support binding
        return null;
    }

    // --- Helper to create the notification channel on Android O+ ---
    private void createNotificationChannel() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            CharSequence name = "App Monitor";
            String description = "Channel for foreground app-monitoring service";
            int importance = NotificationManager.IMPORTANCE_LOW;

            NotificationChannel channel =
                    new NotificationChannel(CHANNEL_ID, name, importance);
            channel.setDescription(description);

            NotificationManager nm =
                    (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);
            nm.createNotificationChannel(channel);
        }
    }
}

File: android/app/src/main/java/com/focusapp/AppMonitorService.kt
package com.focusapp

import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.Service
import android.app.usage.UsageEvents
import android.app.usage.UsageStatsManager
import android.content.Context
import android.content.Intent
import android.os.Build
import android.os.IBinder
import android.util.Log
import com.facebook.react.ReactApplication
import com.facebook.react.ReactInstanceManager
import com.facebook.react.bridge.ReactContext
import com.facebook.react.modules.core.DeviceEventManagerModule

class AppMonitorService : Service() {
    private lateinit var usageStatsManager: UsageStatsManager
    private var monitoringThread: Thread? = null
    private var shouldMonitor = true

    override fun onCreate() {
        super.onCreate()
        usageStatsManager = getSystemService(Context.USAGE_STATS_SERVICE) as UsageStatsManager
        startForegroundService()
    }

    private fun startForegroundService() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            // Create a notification and start as foreground service
            val notification = createNotification()
            startForeground(1, notification)
        }
        startMonitoring()
    }

    private fun createNotification(): android.app.Notification {
        // Create a notification for Android O and above
        val channelId = "app_monitor_channel"
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val channel = android.app.NotificationChannel(
                channelId,
                "App Monitor",
                android.app.NotificationManager.IMPORTANCE_LOW
            )
            val manager = getSystemService(android.app.NotificationManager::class.java)
            manager.createNotificationChannel(channel)
        }

        return android.app.Notification.Builder(this, channelId)
            .setContentTitle("FocusApp Running")
            .setContentText("Monitoring app usage")
            .setSmallIcon(R.mipmap.ic_launcher)
            .build()
    }

    private fun startMonitoring() {
        shouldMonitor = true
        monitoringThread = Thread {
            while (shouldMonitor) {
                try {
                    val foregroundApp = getForegroundApp()
                    foregroundApp?.let { sendAppEvent(it) }
                    Thread.sleep(1000)
                } catch (e: InterruptedException) {
                    Log.e("AppMonitor", "Monitoring interrupted", e)
                }
            }
        }
        monitoringThread?.start()
    }

    private fun getForegroundApp(): String? {
        val endTime = System.currentTimeMillis()
        val startTime = endTime - 1000 * 60 // 1 minute
        val events = usageStatsManager.queryEvents(startTime, endTime)
        val event = UsageEvents.Event()
        var foregroundApp: String? = null
        
        while (events.hasNextEvent()) {
            events.getNextEvent(event)
            if (event.eventType == UsageEvents.Event.MOVE_TO_FOREGROUND) {
                foregroundApp = event.packageName
            }
        }
        
        return foregroundApp
    }

    private fun sendAppEvent(packageName: String) {
        try {
            val reactInstanceManager: ReactInstanceManager? = 
                (application as ReactApplication).reactNativeHost.reactInstanceManager
            val reactContext: ReactContext? = reactInstanceManager?.currentReactContext
            
            reactContext?.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
                ?.emit("APP_IN_FOREGROUND", packageName)
        } catch (e: Exception) {
            Log.e("AppMonitor", "Error sending app event", e)
        }
    }

    override fun onBind(intent: Intent?): IBinder? = null

    override fun onDestroy() {
        shouldMonitor = false
        monitoringThread?.interrupt()
        super.onDestroy()
    }
}